test/extended/apiserver/webhook.go

```diff
@@ -0,0 +1,292 @@
+	})
+})
+
+// annotateNamespace adds the label selector so that the webhook is scoped only to this namespace.
+func annotateNamespace(client kubernetes.Interface, namespace string, labelSelectorKey string) {
+	g.By("annotating the namespace so that the webhook is scoped to this namespace")
+	actualNamespace, err := client.CoreV1().Namespaces().Get(context.Background(), namespace, metav1.GetOptions{})
+	e2e.ExpectNoError(err, "unable to get namespace %s, err %v", namespace, err)
+	if actualNamespace.Labels == nil {
+		actualNamespace.Labels = map[string]string{}
+	}
+	actualNamespace.Labels[labelSelectorKey] = "true"
+	_, err = client.CoreV1().Namespaces().Update(context.Background(), actualNamespace, metav1.UpdateOptions{})
+	e2e.ExpectNoError(err, "unable to update namespace %s, err %v", namespace, err)
+}
+
+// createService creates a service that will route traffic to the webhook server
+// note that a signed serving certificate and the key pair will be created automatically
+// since the service is annotated with service.beta.openshift.io/serving-cert-secret-name
+func createService(client kubernetes.Interface, namespace string, secretName string, serviceName string, servicePort int32, containerPort int32) {
+	g.By("creating a service that will send traffic to the webhook server")
+	serviceLabels := map[string]string{"webhook": "true"}
+	service := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:   namespace,
+			Name:        serviceName,
+			Labels:      map[string]string{"test": "webhook"},
+			Annotations: map[string]string{"service.beta.openshift.io/serving-cert-secret-name": secretName},
+		},
+		Spec: v1.ServiceSpec{
+			Selector: serviceLabels,
+			Ports: []v1.ServicePort{
+				{
+					Protocol:   "TCP",
+					Port:       servicePort,
+					TargetPort: intstr.FromInt(int(containerPort)),
+				},
+			},
+		},
+	}
+	_, err := client.CoreV1().Services(namespace).Create(context.Background(), service, metav1.CreateOptions{})
+	e2e.ExpectNoError(err, "creating service %s in namespace %s", serviceName, namespace)
+}
+
+// deployWebhookServer deploys the webhook server by creating a deployment and verifying the wiring
+func deployWebhookServer(client kubernetes.Interface, image string, namespace string, deploymentName string, secretName string, serviceName string, containerPort int32) {
+	g.By("deploying the webhook server")
+
+	podLabels := map[string]string{"app": "sample-webhook", "webhook": "true"}
+	replicas := int32(1)
+	zero := int64(0)
+	mounts := []v1.VolumeMount{
+		{
+			Name:      "webhook-certs",
+			ReadOnly:  true,
+			MountPath: "/webhook.local.config/certificates",
+		},
+	}
+	volumes := []v1.Volume{
+		{
+			Name: "webhook-certs",
+			VolumeSource: v1.VolumeSource{
+				Secret: &v1.SecretVolumeSource{SecretName: secretName},
+			},
+		},
+	}
+	containers := []v1.Container{
+		{
+			Name:         "sample-webhook",
+			VolumeMounts: mounts,
+			Args: []string{
+				"webhook",
```

Comment:
what exactly is this doing?

Reply:
it runs https://github.com/kubernetes/kubernetes/blob/master/test/images/agnhost/webhook/main.go#L211

the app exposes an API, we use `http.HandleFunc("/always-deny", serveAlwaysDeny)`